src/server/pkg/storage/fileset/storage.go

```diff
@@ -107,194 +99,148 @@ func (s *Storage) newReader(fileSet string, opts ...index.Option) *Reader {
 
 // Open opens a file set for reading.
 // TODO: It might make sense to have some of the file set transforms as functional options here.
-func (s *Storage) Open(ctx context.Context, fileSets []string, opts ...index.Option) (FileSet, error) {
+func (s *Storage) Open(ctx context.Context, ids []ID, opts ...index.Option) (FileSet, error) {
 	var fss []FileSet
-	for _, fileSet := range fileSets {
-		if err := s.store.Walk(ctx, fileSet, func(p string) error {
-			fss = append(fss, s.newReader(p, opts...))
-			return nil
-		}); err != nil {
+	for _, id := range ids {
+		md, err := s.store.Get(ctx, id)
+		if err != nil {
 			return nil, err
 		}
+		switch x := md.Value.(type) {
+		case *Metadata_Primitive:
+			fss = append(fss, s.newReader(id, opts...))
+		case *Metadata_Composite:
+			fs, err := s.Open(ctx, x.Composite.Layers)
```

Comment:
I think we should flatten here. Nesting merge readers might just work, but I am not too sure about it and I think we might be better off having the same simplification that we have for composite filesets, which is no nesting.

Reply:
Composite filesets do allow nesting.  During compaction we flatten before the merge.  In practice we do flatten everything that gets stored long term.